### PR TITLE
Add a test documenting the current state of C++ exception handling

### PR DESCRIPTION
**Pull request by martinboehme**
Swift currently allows C++ exceptions to propagate across Swift stack frames, which is wrong; this test documents that, but also the current behaviors that are fine.

See also the discussion here:
https://forums.swift.org/t/handling-c-exceptions/34823

I'm not sure whether this test will on all platforms, particularly Windows. I may restrict it to certain platforms if build bots show it doesn't work everywhere.

### CODE REVIEW

File: test/ClangImporter/cxx-exceptions-executable.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+var CxxExceptionsTestSuite = TestSuite("CxxExceptions")
+
+// Uncaught C++ exceptions terminate the program.
+CxxExceptionsTestSuite.test("UncaughtException") {
+  expectCrashLater(withMessage: "terminate called")
+  throwException()
+}
+
+// Exceptions can be thrown and caught within C++ code if they don't cross any
+// interop boundaries.
```
Comment (gribozavr) [inline test/ClangImporter/cxx-exceptions-executable.swift]: ... furthermore, Swift can codegen throw and try-catch in a C++ inline functions successfully.

File: test/ClangImporter/cxx-exceptions-executable.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// RUN: %target-build-swift %s -I %S/Inputs/custom-modules/ -o %t/cxx_exceptions -Xfrontend -enable-cxx-interop
+// RUN: %target-codesign %t/cxx_exceptions
+// RUN: %target-run %t/cxx_exceptions
+
+import CxxExceptions
+import StdlibUnittest
+
+var CxxExceptionsTestSuite = TestSuite("CxxExceptions")
+
+// Uncaught C++ exceptions terminate the program.
```
Comment (gribozavr) [inline test/ClangImporter/cxx-exceptions-executable.swift]: I'm not sure what this test shows -- it unwinds through a Swift frame, which we want to disallow.

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

File: test/ClangImporter/cxx-exceptions-executable.swift (hunk 2 above)
Comment (martinboehme) [reply test/ClangImporter/cxx-exceptions-executable.swift]: Maybe that's so, but this test can't "see" this.

The behavior that the test checks for (program is terminated if a C++ exception is not caught) will remain the same even when we've fixed Swift so it doesn't allow unwinding through Swift stack frames.

File: test/ClangImporter/cxx-exceptions-executable.swift (hunk 1 above)
Comment (martinboehme) [reply test/ClangImporter/cxx-exceptions-executable.swift]: Extended the comment.

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Comment by gribozavr** [comment]
@swift-ci Please smoke test Windows

**Comment by martinboehme** [comment]
I was hoping that linking `libstdc++` explicitly in the module map would make this test work at least on macOS as well as Linux, but that doesn't seem to be the case.

So I think I'll have to do the right fix first, which is to make Swift link against the correct library for each platform when `-enable-cxx-interop` is specified.

Until that's done, I'm not looking for further review on this PR. What is the right way to do this -- just leave this PR open with this comment? (If I close it and reopen a new PR, the previous discussion will be lost, which seems undesirable.)

**Comment by gribozavr** [comment]
We can leave it open while it is blocked. Just make sure to close it should plans change and you decide to take another approach.

**Comment by martinboehme** [comment]
Thanks, will do!

**Comment by compnerd** [comment]
Also note that there is a third C++ library that you need to be aware of: msvcprt.

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by martinboehme** [comment]
> Also note that there is a third C++ library that you need to be aware of: `msvcprt`.

Thanks for the heads-up!

As I mention [above](https://github.com/apple/swift/pull/30674#issuecomment-605796131), I've come to the conclusion that it's really not a sustainable approach to link to all the platform-specific libraries locally in this test. Instead, Swift should generally link against the right C++ libraries when `-enable-cxx-interop` is enabled:

https://bugs.swift.org/browse/SR-12469

I'm pausing this PR until that bug is fixed.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**